cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -514,22 +488,21 @@ private long processMetrics() {
               }
 
               if (currentMessageId != null) {
-                if (persistMetaInfo == null) {
-                  persistMetaInfo = new MetricsProcessorStats(currentMessageId);
-                }
-                // update messageId and timestamp information
-                persistMetaInfo.updateStats(currentMessageId, lastMetricTimeSecs);
+                localTopicProcessMeta.updateTopicProcessingStats(currentMessageId, lastMetricTimeSecs);
               }
             } catch (IOException e) {
               LOG.warn("Failed to decode message to MetricValue. Skipped. {}", e.getMessage());
             }
           }
         }
 
-        if (persistMetaInfo != null) {
-          topicMessageIds.put(topicIdMetaKey, persistMetaInfo);
+        if (currentMessageId != null) {
+          // update the last processed timestamp in local topic meta and update the topicProcessMetaMap with this
+          // local topic meta for the topic
+          localTopicProcessMeta.updateLastProcessedTimestamp();
+          topicProcessMetaMap.put(topicIdMetaKey, localTopicProcessMeta);
         }
-
+        // TODO q: should we try persist only if current messageId is non-null(have processed)? currently we always try
```

Comment:
if it is null, then we have not made any changes, right? But it does not harm, and the existing code does it, too. 

There is at least one case where it is useful. Suppose you have only 2 threads. Thread1 processes some metrics and persists. Thread2 also processes metrics, but does not persist because Thread1 has the lock. But Thread1 does not see Thread2's changes, so they don't get persisted. In the next round, both threads might not process any metrics. Then these changes from Thread2 would not be persisted, only when one thread processes something again. 

If you always attempt to persist, you are sure that changes get persisted quickly.